drivers/sdhc/rcar_mmc.c

```diff
@@ -528,10 +540,34 @@ static int rcar_mmc_dma_rx_tx_data(const struct device *dev,
 	rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_ADDR_L, dma_addr);
 	rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_ADDR_H, 0);
 
+#ifdef CONFIG_RCAR_MMC_DMA_IRQ_DRIVEN_SUPPORT
+	rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_INFO2_MASK,
+		(uint32_t)(is_read ? (~RCAR_MMC_DMA_INFO2_ERR_RD) :
+				     (~RCAR_MMC_DMA_INFO2_ERR_WR)));
+
+	reg = rcar_mmc_read_reg32(dev, RCAR_MMC_DMA_INFO1_MASK);
+	reg &= ~dma_info1_poll_flag;
+	rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_INFO1_MASK, reg);
 	rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_CTL, RCAR_MMC_DMA_CTL_START);
 
+	ret = k_sem_take(&dev_data->irq_xref_fin,
```

Comment:
I wonder what will happen if you'll get of sync in you semaphore usage. For example, there will be an extra DMA interrupt, of one of the interrupts will be late, so you'll return with timeout from ` k_sem_take`, but interrupt handler then will increase semaphore count. In the next operation, you'll return from `k_sem_take` immediately, which is wrong.

Linux in this case use "completions", which are free from such an issues. Or, in some cases, drivers check status register and restart wait for interrupt if transaction is not ready.

Reply:
good catch!

I have masked interrupts inside IRQ handler and IRQs have been masked before triggering of DMA start, so it isn't possible to have two or more IRQs in a row, but late arriving of the current interrupt is still possible, e.g. k_sem_take returns error code, I think one action will be enough here:
* reset the semaphore after resetting DMA IRQ masks.

Reply:
This is a bit unorthodox, but I can't see reason why this might now work.